features/graph/dao/api/src/main/java/org/opennms/netmgt/topology/EdgeEntity.java

```diff
@@ -28,39 +28,39 @@
 
 package org.opennms.netmgt.topology;
 
-import javax.persistence.CascadeType;
+import java.util.Optional;
+
 import javax.persistence.DiscriminatorValue;
 import javax.persistence.Entity;
-import javax.persistence.JoinColumn;
-import javax.persistence.ManyToOne;
 
 @Entity
 @DiscriminatorValue("edge")
 public class EdgeEntity extends AbstractGraphEntity {
 
-    // TODO MVR we don't need this. This could be a property instead (due to vertex-ref), so source or target may not exist in the table
-    @ManyToOne(optional = false, cascade = CascadeType.ALL)
-    @JoinColumn(name="source_vertex_id", referencedColumnName = "id", nullable = false, updatable = true)
-    private VertexEntity source;
+    private final static String PROPERTY_SOURCE = "sourceVertexRef";
+    private final static String PROPERTY_TARGET = "targetVertexRef";
+
+    public VertexRef getSource() {
+        return Optional.ofNullable(getPropertyValue(PROPERTY_SOURCE)).map(VertexRef::new).orElse(null);
+    }
 
-    // TODO MVR we don't need this. This could be a property instead (due to vertex-ref)
-    @ManyToOne(optional = false, cascade = CascadeType.ALL)
-    @JoinColumn(name="target_vertex_id", referencedColumnName = "id", nullable = false, updatable = true)
-    private VertexEntity target;
+    public void setSource(String namespace, String id) {
+        setSource(new VertexRef(namespace, id));
+    }
 
-    public VertexEntity getSource() {
-        return source;
+    public void setSource(VertexRef source) {
+        this.setProperty(PROPERTY_SOURCE, String.class, source.toString());
```

Comment:
Wouldn't make `asStringRepresentation()` more sense instead of hijacking `toString()` for this purpose?
Maybe even allow `VertexRef` as supported property values instead.

Reply:
agree, lets make it a supported property, then we have it typesafe

Reply:
done, see also comment above class